replica/tablets.cc

```diff
@@ -308,4 +310,272 @@ future<std::vector<canonical_mutation>> read_tablet_mutations(seastar::sharded<r
     co_return std::move(result);
 }
 
+// This sstable set provides access to all the stables in the table, using a snapshot of all
+// its tablets/storage_groups compound_sstable_set:s.
+// The managed sets cannot be modified through tablet_sstable_set, but only jointly read from, so insert() and erase() are disabled.
+class tablet_sstable_set : public sstables::sstable_set_impl {
+    schema_ptr _schema;
+    size_t _log2;
+    // Keep a single (compound) sstable_set per tablet/storage_group
+    std::vector<lw_shared_ptr<sstables::sstable_set>> _sstable_sets;
+    size_t _size;
+    uint64_t _bytes_on_disk;
+
+public:
+    explicit tablet_sstable_set(const storage_group_manager& sgm) noexcept
+        : _schema(sgm.table().schema())
+        , _log2(sgm.log2_storage_groups())
+    {
+        _sstable_sets.reserve(sgm.storage_groups().size());
+        for (const auto& sg : sgm.storage_groups()) {
```

Comment:
This isn't going to work well. Consider a 1 PB table, one table in the cluster. It will have 200,000 tablets. Of these only 100 will be on this shard. We'll iterate over 200,000 items to find 100.

Reply:
The storage groups thing needs to maintain multiple data structures. A vector for fast lookup. A std::flat_map for in-order iteration.

Reply:
Agreed, but it's orthogonal to this series since it's a preexisting problem. I am planning to fix it with this, see: https://github.com/raphaelsc/scylla/blob/ceaed7a842ec510db36b9d0119d010087da58565/utils/chunked_vector.hh#L502

Reply:
ok